### DISCUSSION TITLE
Links as a content type (.url) - opens as a file rather than a link; no preview image available

### DISCUSSION THREAD

**Discussion by popeadam** [microsoft-search/pnp-modern-search#3036] [Ideas]
**Is your feature request related to a problem? Please describe.**
**Version used**
4.3.0

In a site pages library we have added a number of 'Links' to other pages. When clicked on from the search results they are opening as a file, where you then need to click on the 'Open' button. They also appear with no preview image.

It would be great if, when clicking on the links, they opened the URL rather than the file, and we could specify a preview image that gets picked up by the AutoPreviewImageURL slot.

**To Reproduce**
1. Go to site pages library and click New/Link
2. Go to search results web part
3. Note it has no thumbnail - though we've tried specifying it in the SiteBannerURL column
4. Click on a link
5. Note that it opens as a document rather than opening the URL associated with the link as it does in the library

**Describe the solution you'd like**
When clicking on a link, it should open the URL specified in the link rather than the file 

**Screenshots**
Adding a link:
![image](https://user-images.githubusercontent.com/54813914/151196350-6b6ab40a-c537-4a76-bbd8-38018b9f20b9.png)

Links with no thumbnails:
![image](https://user-images.githubusercontent.com/54813914/151197141-81cb7c04-bce0-4f58-90af-f4d48e20c27d.png)

Clicking on a link from search results takes me to the file:
![image](https://user-images.githubusercontent.com/54813914/151196601-900055e0-2dcd-45a4-91d1-296344fd187a.png)

**Desktop (please complete the following information):**
 - Browser: Chrome

**Comment by wobba** [comment]
Tagging this as an enhancement. You can fix it yourself now using the result types feature where you trigger on the content type and specify the template and properties to use.

**Comment by jameelrrs** [comment]
@wobba , Can you help with the property for getting URL field for .url files. Earlier in V3, We could see a pre-defined property "ShortcutUrl".  Some how this is not appearing in V4 version.  Is the name changed or we are missing anything here.  Please guide

**Comment by wobba** [comment]
@jameelrrs there is no property, you have to read the contents of the file with a separate API call.

**Comment by SjoerdV** [comment]
I used this 'column template' as a workaround.

```html
{{#compare FileType '!=' 'url' }}
<a href="{{slot item @root.slots.PreviewUrl}}" target="_blank" style="color: {{@root.theme.semanticColors.link}}">
	{{slot item @root.slots.Title}}
</a>
{{else}}
<a href="{{itemAt (split (itemAt (split HitHighlightedSummary 'URL=') 1) '<') 0}}" target="_blank" style="color: {{@root.theme.semanticColors.link}}">
	{{slot item @root.slots.Title}}
</a>
{{/compare}}
```

Key being the ```{{itemAt (split (itemAt (split HitHighlightedSummary 'URL=') 1) '<') 0}}``` part. Seems like the HitHighlightedSummary property includes the 'ShortcutUrl' data now

NOTE: this seems not to work (at least in version 4.4.1) when your Query Template uses a token like {searchTerms} as mentioned here: #2010

**Reply by spinyriptide** [reply]
THIS WORKS! I was able to adapt this snippet to the Managed Card "Tags" field on config panel 2 to at least get the Title and URL to work. It's important to add  data-interception="off" after target="_blank" so the link opens in a new window. 

I'm now looking at how to display the Description field below it where I display Summary for other content.

**Reply by simoncumming** [reply]
This method almost worked for me, but the URLs I had seem like they are too long, so the links that would result weren't functional because they were missing the final characters.

Solution from @zkent  below worked for me though.

**Comment by wobba** [comment]
@SjoerdV Relying on `HitHighlightedSummary` for data is likely a very brittle solution. Implementing the code from v3 is the only 100% solution.

**Comment by SjoerdV** [comment]
Ok thanks, than I'll gladly wait for a more robust solution to be ported to v4.

In this case having something that seems to work is better than nothing at all. And how likely is it the value for the HitHighlightedSummary property of a Link contenttype result will change any time soon... I am feeling lucky today ;-)

**Comment by wobba** [comment]
Unlikely, as the the item itself has no real data.

**Comment by zkent** [comment]
Has anyone figured this out?  I have replaced `AutoPreviewUrl` with `ShortcutUrl` in the Layout Slots configuration and it works great in most instances.  However, it appears that `ShortcutUrl` contains the URL _twice_.  This breaks other instances where the second instance of the URL appears as an unexpected URL parameter of the first URL.

**Comment by wobba** [comment]
@zkent You can probably use handlebar helper functions to split the data and have one URL only.

**Comment by zkent** [comment]
I am fairly new to PnP and handlebars.  Can I get some tips as to where I would add this code?

EDIT:  I figured it out.  I created a slot named `ShortcutUrl` to hold the value.  I then replaced this

```html
<a href="{{slot item @root.slots.PreviewUrl}}" target="_blank" style="color:{{@root.theme.semanticColors.link}}" data-interception="off" rel="noopener noreferrer">{{slot item @root.slots.Title}}
``` 
with
```html
{{#with (split (slot item @root.slots.ShortcutUrl) ',')}}
    <a href="{{[0]}}" target="_blank" data-interception="off" style="color:{{@root.theme.semanticColors.link}}" rel="noopener noreferrer">{{../Title}}</a>
{{/with}}
``` 

For some reason `{{slot item @root.slots.Title}}` did not render inside the `with` helper so I had to use `{{../Title}}`

**Reply by spinyriptide** [reply]
Bless you! I'm looking for a solution to this since the URL content type is HUGELY important to anchor SharePoint as a comprehensive knowledge system. It appears in every library, which also makes it very versatile. It works fine with out-of-box SharePoint web parts -- the link opens without a hitch. With PnP v4 Search Results, it has all the faults mentioned. Worse, a browser security warning accompanies a user click about downloading and opening such files. This is definitely a priority enhancement in my view. Otherwise, the v4 results web  part is absolutely liberating and brilliant.

**Reply by spinyriptide** [reply]
I struggled with this one after all. It doesn't parse the URL from the summary fully. @SjoerdV shared something that does, above.

**Reply by simoncumming** [reply]
Thanks so much for this, worked really well.

Just FYI for anyone else very new to this `ShortcutUrl` is a property that needs to be added in 'Selected Properties' before you can add it to a layout slot. After that it seems to work for me perfectly.

**Reply by simoncumming** [reply]
Ended up combining this with the code from @SjoerdV so it could handle Urls and Non-Urls

This is the handlebars that I have put in the 'Title' section of 'Manage Columns'

`{{#compare FileType '!=' 'url' }}
<a href="{{slot item @root.slots.PreviewUrl}}" target="_blank" style="color: {{@root.theme.semanticColors.link}}">
	{{slot item @root.slots.Title}}
</a>
{{else}}
{{#with (split (slot item @root.slots.ShortcutUrl) ',')}}
    <a href="{{[0]}}" target="_blank" data-interception="off" style="color:{{@root.theme.semanticColors.link}}" rel="noopener noreferrer">{{../Title}}</a>
{{/with}}
{{/compare}}`

**Reply by popeadam** [reply]
The above solutions didn't work using a card layout - but the suggestions certainly helped! What did:

1. On tab 4 under 'Edit Properties' scroll down to the bottom of 'selected properties' and add "ShortcutUrl"
2. Republish the page, edit it, and on tab 1 under Layout slots click 'Customise' 
3. Add a new ShortcutUrl slot mapped to the ShortcutUrl property you added in step 1
4. On tab 2 click 'Managed Card fields' or 'Manage Columns' depending on the layout you're using
5. If you're using a details layout, to the right of the Title column click on the handlebars editor and use the code above in the Title field
6. If you're using a card layout, to the right of the Url column click on the handlebars editor and paste in the following:

```
{{#compare FileType '!=' 'url'}}
 {{slot item @root.slots.PreviewUrl}} 
{{else}}
  {{#with (split (slot item @root.slots.ShortcutUrl) ',')}}
    {{[0]}}
  {{/with}}
{{/compare}}
```

**Comment by wobba** [comment]
I'll leave this open and see if we can get this logic in place for these URL's without handlebar helper changes.

**Comment by spinyriptide** [comment]
Now that it's 2024, curious when an enhancement to this will be in place. In SharePoint, Link and Repost behave differently.  As my workaround, I'm using Repost and de-promoting items in Site Pages with the Promoted State column and a Handlebars toggle behavior. However, Link is more ubiquitous in lists and libraries and doesn't inherit the News promoted state in Pages, so it does not require this loop de loop. Leveraging URLs in general to other ecosystem resources is a key way to anchor a complex ecosystem in SharePoint using these dynamic web parts. My fingers are crossed.

**Reply by wobba** [reply]
When a developer with time on their hands has time and sees a need, that is when changes happen. The nature of a community project.
